/*
 * SPDX-FileCopyrightText: Copyright (c) 2023-2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

syntax = "proto3";

package holoscan.distributed;

import "system_resource.proto";
import "result.proto";

message FragmentAllocationRequest {
  string worker_ip = 1;
  string worker_port = 2;
  repeated string fragment_names = 3;
  AvailableSystemResource available_system_resource = 4;
}

message FragmentAllocationResponse {
  Result result = 1;
}

message WorkerExecutionFinishedRequest {
  string worker_ip = 1;
  string worker_port = 2;
  Result status = 3;
}

message WorkerExecutionFinishedResponse {
  Result result = 1;
}

message InitiateShutdownRequest {
  string fragment_name = 1;  // Name of fragment requesting shutdown
}

message InitiateShutdownResponse {
  Result result = 1;
}

service AppDriverService {
  rpc AllocateFragments(FragmentAllocationRequest) returns (FragmentAllocationResponse) {}
  rpc ReportWorkerExecutionFinished(WorkerExecutionFinishedRequest) returns (WorkerExecutionFinishedResponse) {}
  rpc InitiateShutdown(InitiateShutdownRequest) returns (InitiateShutdownResponse) {}
}
